lib/grammar/rules/command_name.cpp

```diff
@@ -0,0 +1,51 @@
+#include "shell/grammar/rules/command_name.hpp"
+#include "shell/grammar/rules/assignment_word.hpp"
+#include "shell/assert.hpp"
+#include "shell/grammar/grammar_util.hpp"
+
+namespace shell {
+
+Rule CommandName::make() {
+	return Rule {
+	    CommandName::options,
+	};
+}
+
+vector<Rule::Option> CommandName::options() {
+	// TODO: implement
+	return {Rule::Terminal {handler}};
+}
+
+optional<Ast::Node> CommandName::handler(TokenProvider &tokens) {
+	auto token = tokens.consumeIf([](Token &token) { return token.getType() == Token::Type::Token; });
+	if (!token.has_value()) {
+		return nullopt;
+	}
+	// Rule 7a
+	auto &word_token = token->get<WordToken>();
+	auto &value = word_token.value;
+	if (value.find('=') == std::string::npos) {
+		// Rule 1
+		auto reserved_word = Token::exactReservedWordType(value);
+		if (!reserved_word.has_value()) {
+			return Ast::Literal(Token(Token::Type::Word, WordToken {word_token.value}));
+		}
+		// WORD is requested here
+		return nullopt;
+	}
+	// Rule 7b
+	D_ASSERT(!value.empty());
+	if (value[0] == '=') {
+		return Ast::Literal(Token(Token::Type::Word, WordToken {word_token.value}));
+	}
+	auto equals_pos = value.find('=');
+	D_ASSERT(equals_pos != std::string::npos);
+	auto maybe_name = value.substr(0, equals_pos);
+	if (AssignmentWord::isValidName(maybe_name)) {
+		// WORD is requested here
+		return nullopt;
+	}
+	return Ast::Literal(Token(Token::Type::Word, WordToken {word_token.value}));
```

Comment:
So in the standard it says here that it is unspecified which one is returned, I suppose that means it can be both, so in the case that we expect a Word we can return a Word? Am I interpreting that correctly?

Reply:
Yea I'm not entirely happy how I'm handling word/assignment-word/name specification currently, a lot of duplicated code

Reply:
I also think that's what it means, but the standard is quite vague

Reply:
(I'm guessing the code it used to base the standard on was less than ideal)

Reply:
If the standard was written at the hand of the code then that explains a lot haha

Reply:
Let's just assume for now that that's the case